util/bloom_impl.h

```diff
@@ -0,0 +1,140 @@
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Implementation details of various Bloom filter implementations used in
+// RocksDB. (DynamicBloom is in a separate file for now because it
+// supports concurrent write.)
+
+#pragma once
+#include <stddef.h>
+#include <stdint.h>
+
+#include "rocksdb/slice.h"
+
+namespace rocksdb {
+
+// A legacy Bloom filter implementation with no locality of probes (slow).
+// It uses double hashing to generate a sequence of hash values.
+// Asymptotic analysis is in [Kirsch,Mitzenmacher 2006], but known to have
+// subtle accuracy flaws for practical sizes [Dillinger,Manolios 2004].
+//
+// DO NOT REUSE - faster and more predictably accurate implementations
+// are available at
+// https://github.com/pdillinger/wormhashing/blob/master/bloom_simulation_tests/foo.cc
+// See e.g. RocksDB DynamicBloom.
+//
+class LegacyNoLocalityBloomImpl {
+public:
+  static inline void AddHash(uint32_t h, uint32_t total_bits,
+                             int num_probes, char *data) {
+    const uint32_t delta = (h >> 17) | (h << 15);  // Rotate right 17 bits
+    for (int i = 0; i < num_probes; i++) {
+      const uint32_t bitpos = h % total_bits;
+      data[bitpos/8] |= (1 << (bitpos % 8));
+      h += delta;
+    }
+  }
+
+  static inline bool HashMayMatch(uint32_t h, uint32_t total_bits,
+                                  int num_probes, const char *data) {
+    const uint32_t delta = (h >> 17) | (h << 15);  // Rotate right 17 bits
+    for (int i = 0; i < num_probes; i++) {
+      const uint32_t bitpos = h % total_bits;
+      if ((data[bitpos/8] & (1 << (bitpos % 8))) == 0) {
+        return false;
+      }
+      h += delta;
+    }
+    return true;
+  }
+};
+
+
+// A legacy Bloom filter implementation with probes local to a single
+// cache line (fast). Because SST files might be transported between
+// platforms, the cache line size is a parameter rather than hard coded.
+// (But if specified as a constant parameter, an optimizing compiler
+// should take advantage of that.)
+//
+// When ExtraRotates is false, this implementation is notably deficient in
+// accuracy. Specifically, it uses double hashing with a 1/512 chance of the
+// increment being zero (when cache line size is 512 bits). Thus, there's a
+// 1/512 chance of probing only one index, which we'd expect to incur about
+// a 1/2 * 1/512 or absolute 0.1% FP rate penalty. More detail at
+// https://github.com/facebook/rocksdb/issues/4120
+//
+// DO NOT REUSE - faster and more predictably accurate implementations
+// are available at
+// https://github.com/pdillinger/wormhashing/blob/master/bloom_simulation_tests/foo.cc
+// See e.g. RocksDB DynamicBloom.
+//
+template <bool ExtraRotates>
+class LegacyLocalityBloomImpl {
+private:
+  static inline uint32_t GetLine(uint32_t h, uint32_t num_lines) {
+    uint32_t offset_h = ExtraRotates ? (h >> 11) | (h << 21) : h;
```

Comment:
I'm not expert on template and this is the first time I see template type as a boolean, which is put into branch. I guess the expectation is that, the compiler will remove the branching in compile time. Is my understanding correct?

Reply:
That's right. Although a compiler could opt for smaller code size by making it a runtime check, that would be unusual and "inline" is a strong hint not to.

Let me do a quick perf test to be extra sure...

Reply:
The only thing I'm seeing is small to significant improvement in speed. Not sure what I did, but I did get rid of some haphazard use of size_t where we already know uint32_t suffices (for better or worse), some field references vs. local variable references, and I don't know.